arbiter/drivers/onedrive.cpp

```diff
@@ -0,0 +1,292 @@
+    //next set of items
+    do
+    {
+        http::Query queries;
+        auto res(https.internalGet(resource.childrenEndpoint(), headers, queries));
+        const json obj(json::parse(res.data()));
+        const json parsedQueries(json::parse(getQueries(pageUrl)));
+
+        for (auto& it: parsedQueries.items())
+            queries[it.key()] = it.value();
+
+        if (!obj.contains("value") || !res.ok())
+        {
+            std::cout << "Could not get OneDrive item  " << resource.childrenEndpoint()
+                << " with response code " << res.code() << std::endl;
+            return std::vector<std::string>();
+        }
+
+        //parse list
+        const json list(obj.at("value"));
+        for (auto& i: list.items())
+        {
+            const auto data(i.value());
+            const std::string fileName(data.at("name").get<std::string>());
+            const std::string filePath(path + "/" + fileName);
+
+            result.push_back(filePath);
+            if (data.contains("folder") && recursive)
+            {
+                //restart process with new file head
+                const auto children(processList(filePath, recursive));
+
+                //add result of children processes to the parent
+                result.insert(result.end(), children.begin(), children.end());
+            }
+        }
+
+        //check for link to next set
+        if (obj.contains("@odata.nextLink"))
+            pageUrl = obj.at("@odata.nextLink");
+        else
+            break;
+
+    } while (true);
+
+    return result;
+}
+
+std::vector<std::string> OneDrive::glob(std::string path, bool verbose) const
+{
+    path.pop_back();
+    bool recursive(path.back() == '*');
+    if (recursive)
+        path.pop_back();
+
+    if (path.back() == '/')
+        path.pop_back();
+
+    return processList(path, recursive);
+}
+
+bool OneDrive::get(
+        const std::string path,
+        std::vector<char>& data,
+        const http::Headers userHeaders,
+        const http::Query query) const
+{
+    Resource resource(path);
+    http::Headers headers(m_auth->headers());
+    headers["Content-Type"] = "application/octet-stream";
+    headers.insert(userHeaders.begin(), userHeaders.end());
+
+    drivers::Https https(m_pool);
+    const auto res(https.internalGet(resource.binaryEndpoint(), headers));
+
+    if (!res.ok()) {
+        std::cout <<
+            "Failed get - " << res.code() << ": " << res.str() << std::endl;
+        return false;
+    }
+
+    data = res.data();
+    return true;
+}
+
+OneDrive::Auth::Auth(std::string s) {
+    const auto config = json::parse(s);
+    m_token = config.at("access_token").get<std::string>();
+    m_refresh = config.at("refresh_token").get<std::string>();
+    m_redirect = config.at("redirect_uri").get<std::string>();
+    m_id = config.at("client_id").get<std::string>();
+    m_secret = config.at("client_secret").get<std::string>();
+    m_expiration = Time().asUnix();
```

Comment:
On startup we're using the current time as expiration to immediately refresh, is that correct?  I think that intent would be more clear as the initial state by initializing `m_expiration` to `0` instead of the current time (do it in the header and remove the assignment from here).